### PR TITLE
backend support for Facebook/LinkedIn JS API login from extension

### PR DESCRIPTION
**Pull request by 2is10**
@raymondng I factored out some of the code from `MobileAuthController` to `AuthCommander` for reuse by `ExtAuthController`. To vet (validate) tokens, `ExtAuthController` just verifies their accompanying signatures.

Both Facebook’s and LinkedIn’s JS APIs return tokens with signatures. Facebook’s token is a short-lived OAuth 2 token (works with their REST API). LinkedIn’s token is not a valid OAuth 2 token, but it can be exchanged for a valid OAuth 1.0a token. In my opinion, in neither case is what we have compelling enough to justify making a request to the provider’s REST API or persisting a new token to our database.

### CODE REVIEW

**Comment by 2is10** [comment]
I’m still going to do some minor cleanup and add some tests, but I’d appreciate a review of the flow. Come to my desk if you’d like to see it working.

File: kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala [hunk 1]
```diff
@@ -239,4 +240,26 @@ class AuthCommander @Inject()(
...
+        authenticator =>
+          Ok(Json.obj("code" -> "user_logged_in", "sessionId" -> authenticator.id))
+            .withSession(newSession - SecureSocial.OriginalUrlKey - IdentityProvider.SessionId - OAuth1Provider.CacheKey)
+            .withCookies(authenticator.toCookie)
+      )
+    } getOrElse {
+      log.info(s"[loginWithTrustedSocialIdentity($identityId})] no kifi user")
+      NotFound(Json.obj("error" -> "user_not_found"))
+    }
+  }
```
Comment (2is10) [inline kifi-backend/modules/shoebox/app/com/keepit/commanders/AuthCommander.scala:262]: (all of this code is just moved from `MobileAuthController`)

File: kifi-backend/modules/shoebox/app/com/keepit/common/social/FacebookSocialGraph.scala [hunk 2]
```diff
@@ -163,6 +171,34 @@ class FacebookSocialGraph @Inject() (
     (json \ "gender").asOpt[String].map(Gender.key -> Gender(_).toString)
   ).flatten.toMap
 
+  def vetJsAccessToken(provider: IdentityProvider, json: JsValue): Future[IdentityId] = {
```
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/social/FacebookSocialGraph.scala]: Nice :+1:

File: kifi-backend/modules/shoebox/app/com/keepit/common/social/LinkedInSocialGraph.scala [hunk 3]
```diff
@@ -127,6 +154,73 @@ class LinkedInSocialGraph @Inject() (
...
+    val signatureOrder = (json \ "signature_order").as[Seq[String]]
+    val signatureBase = signatureOrder.fold(""){ case (b, k) => b + (json \ k).as[String] }
+    val mac = Mac.getInstance("HmacSHA1")
+    mac.init(new SecretKeySpec(OAuth.percentEncode(settings.clientSecret).getBytes(OAuth.ENCODING), "HmacSHA1"))
+    val computedSignature = mac.doFinal(signatureBase.getBytes(OAuth.ENCODING))
+    if (!java.util.Arrays.equals(new Base64().decode(signature), computedSignature)) {
+      return Future.failed(new OAuthExpectationFailedException("signature mismatch"))
+    }
+
+    // exchange JS API bearer token for OAuth 1.0a access token (LinkedIn does not yet support exchange for 2.0 access token)
```
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/common/social/LinkedInSocialGraph.scala]: LOL LinkedIn does not make things easy. Aside from vendor-specific limitation I don't know if there's any value in persisting (supporting?)  OAuth1.0a

File: kifi-backend/modules/shoebox/app/com/keepit/controllers/mobile/MobileAuthController.scala [hunk 4]
```diff
@@ -91,44 +88,24 @@ class MobileAuthController @Inject() (
...
-      for {
-      provider   <- Registry.providers.get(providerName)
-      oauth2Info <- request.body.asOpt[OAuth2Info]
+    (for {
+      provider <- Registry.providers.get(providerName)
+      oAuth2Info <- request.body.asOpt[OAuth2Info]
     } yield {
       Try {
-        provider.fillProfile(SocialUser(IdentityId("", provider.id), "", "", "", None, None, provider.authMethod, oAuth2Info = Some(oauth2Info))) // get fb socialId
+        provider.fillProfile(SocialUser(IdentityId("", provider.id), "", "", "", None, None, provider.authMethod, oAuth2Info = Some(oAuth2Info)))
```
Comment (raymondng) [inline kifi-backend/modules/shoebox/app/com/keepit/controllers/mobile/MobileAuthController.scala:142]: lgtm

**Comment by 2is10** [comment]
Thanks for reviewing Ray!

**Pull request merged**